mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,416 @@
+}
+
+unsafe impl Send for RingBuffer {}
+unsafe impl Sync for RingBuffer {}
+
+fn create_file(file_name: String) -> IOResult<File> {
+    OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(&PathBuf::from(file_name))
+}
+
+impl RingBuffer {
+    pub fn new(file_name: String, max_file_size: usize, flush_options: FlushOptions) -> Self {
+        let file = create_file(file_name).expect("Failed to create file");
+        file.set_len(max_file_size as u64)
+            .expect("Failed to set file size");
+        let mmap = unsafe { MmapMut::map_mut(&file).expect("Failed to open mmap") };
+        Self {
+            inner: Arc::new(Mutex::new(mmap)),
+            max_file_size,
+            flush_options,
+            pointers: Arc::new(FilePointers::default()),
+            waker: None,
+            flush_state: Arc::new(FlushState::new()),
+        }
+    }
+}
+
+fn should_flush(flush_options: FlushOptions, flush_state: &FlushState) -> bool {
+    match flush_options {
+        FlushOptions::AfterEachWrite => true,
+        FlushOptions::AfterXWrites(xwrites) => flush_state.writes.load(Ordering::SeqCst) >= xwrites,
+        FlushOptions::AfterXBytes(xbytes) => {
+            flush_state.bytes_written.load(Ordering::SeqCst) >= xbytes
+        }
+        FlushOptions::AfterXMilliseconds(xmillis_elapsed) => {
+            flush_state.millis_elapsed.load(Ordering::SeqCst) >= xmillis_elapsed
+        }
+        FlushOptions::Off => false,
+    }
+}
+
+struct RingBufferWriteFuture {
+    pointers: Arc<FilePointers>,
+    flush_options: FlushOptions,
+    mmap: Arc<Mutex<MmapMut>>,
+    flush_state: Arc<FlushState>,
+    value: Vec<u8>,
+    max_file_size: usize,
+}
+
+impl Future for RingBufferWriteFuture {
+    type Output = StorageResult<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let timer = Instant::now();
+        let data = binary_serialize(&self.value)?;
+        let data = Data::new(data);
+        let serialized_data = binary_serialize(&data)?;
+
+        let data_size = serialized_data.len();
+        let data_hash = calculate_hash(&data);
+
+        let write_index = self.pointers.write.load(Ordering::SeqCst);
+        let read_index = self.pointers.read.load(Ordering::SeqCst);
+
+        let block_header = BlockHeaderWithHash::new(data_size, write_index, data_hash);
+        let serialized_block_header = binary_serialize(&block_header)?;
+        let block_size = binary_serialize_size(&BlockHeaderWithHash::new(0, 0, 0))?;
+
+        let total_size = block_size + data_size;
+        // Check to see if we might corrupt data if we write, if so, return pending.
+        // As trying later might be successful.
+        if write_index < read_index {
+            if write_index + total_size > read_index {
+                println!("write pending");
+                return Poll::Pending;
+            }
+        }
+        cx.waker().wake_by_ref();
```

Comment:
I believe waker should not be called from the poll method of the future. Instead you need to "remember" the waker somehow. By "remember" I mean that you need to store it in the struct that implements `StreamWakeableState`. 

Next time a `batch` (or rather `remove`) method is called on the `StreamWakeableState`, at the end of the reading, you would wake a waker if there is one.

Reply:
Ahh okay, so wake once the condition for blocking has passed